File: Coordinate/mapCoordinate.py
import numpy as np
import cv2
import matplotlib.pyplot as plt

def onclick(event):
    if event.xdata is not None and event.ydata is not None:
        x, y = int(event.xdata), int(event.ydata)
        print(f'Clicked at: ({x}, {y})')
        click_positions.append((x, y))
        
        text_x = x - 60
        text_y = y - 30
        # Display the coordinates on the image
        cv2.circle(img_rgb, (x, y), radius=10, color=(0, 0, 255), thickness=-1)
        cv2.putText(img_rgb, f'{x}, {y}', (text_x, text_y), cv2.FONT_HERSHEY_SIMPLEX, 1, (0, 0, 255), 2)
        plt.imshow(img_rgb)
        plt.axis('on')
        plt.show()

if __name__ == "__main__":
    # Path to the image
    filename = "/Users/yaojiamin/Documents/SIUE/Course/CS456 Algorithms/project/Europe.png"

    img = cv2.imread(filename)

    if img is None:
        print("Error: Unable to load image.")
    else:
        # Convert the image from BGR (OpenCV format) to RGB (Matplotlib format)
        img_rgb = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)

        # Global variable to store click positions
        click_positions = []

        # Display the image with Matplotlib and connect the click event
        fig, ax = plt.subplots()
        ax.imshow(img_rgb)
        cid = fig.canvas.mpl_connect('button_press_event', onclick)
        plt.axis('on')
        plt.show()

File: Coordinate/mapCoordinate.cpp
#include <opencv2/opencv.hpp>
#include <iostream>
#include <vector>

// Global variables
cv::Mat img, img_display;
std::vector<cv::Point> click_positions;

// Mouse callback function
void onMouse(int event, int x, int y, int, void*)
{
    if (event == cv::EVENT_LBUTTONDOWN)
    {
        // Record the clicked position
        cv::Point clicked_point(x, y);
        click_positions.push_back(clicked_point);

        // Display the coordinates in the console
        std::cout << "Clicked at: (" << x << ", " << y << ")" << std::endl;

        // Annotate the image with a circle and the coordinates
        cv::circle(img_display, clicked_point, 10, cv::Scalar(0, 0, 255), -1); // Red circle
        cv::putText(img_display, std::to_string(x) + ", " + std::to_string(y),
                    cv::Point(x - 60, y - 30), cv::FONT_HERSHEY_SIMPLEX, 1, cv::Scalar(0, 0, 255), 2);

        // Display the updated image
        cv::imshow("Image", img_display);
    }
}

int main()
{
    // Path to the image
    std::string filename = "../Image/Europe.png";

    // Load the image
    img = cv::imread(filename);

    if (img.empty())
    {
        std::cerr << "Error: Unable to load image." << std::endl;
        return -1;
    }

    // Create a copy of the image for displaying annotations
    img_display = img.clone();

    // Display the image in a window
    cv::namedWindow("Image", cv::WINDOW_AUTOSIZE);
    cv::imshow("Image", img_display);

    // Set the mouse callback function
    cv::setMouseCallback("Image", onMouse);

    // Wait indefinitely for a key press
    cv::waitKey(0);

    return 0;
}
